### DISCUSSION TITLE
How to use functions in beeware

### DISCUSSION THREAD

**Discussion by montyonthebonty** [beeware/beeware#164] [General]
Good evening

Quite new to Beeware, but enjoying experimenting with it. However, I have a question, about how to set out the code. I have managed to create a 'build Screen' function to populate the main screen with the various widgets, but I want the button 'btnStart' to be able to manipulate the various widgets, using the function 'getImage'. Where do I put the function and what parameters do I pass in to be able to change the widgets themselves.

I've tried lots of different combinations but can't get it to work. Below is the bit of code I have so far (apologies if it's a bit amateurish!)

Thanks
Chris

```  """
  Route-learning assistant
  """
  import toga
  from toga.style import Pack
  from toga.style.pack import COLUMN, ROW

  import requests
  import json

  class Routemaster(toga.App):

      def startup(self):

          def frChange(widget):
              pass

          def whrChange(widget):
              pass

          def getImage(widget):
              pass

          # Build the main welcome screen
          def buildMainScreen():

              # Create the blank layout
              mainBox = toga.Box(style=Pack(direction=COLUMN, padding=20))

              # Create sublayouts
              locRow = toga.Box(style=Pack(direction=ROW))
              optRow = toga.Box(style=Pack(direction=ROW))
              quizBox = toga.Box(style=Pack(direction=ROW))
              resBox = toga.Box(style=Pack(direction=COLUMN))


              # Create the base widgets
              lblTitle = toga.Label('Routemaster')
              lblIntro = toga.Label('Welcome to the Routemaster Quiz app')
              lblFrom = toga.Label('Between')
              selFrom = toga.Selection(id='selOrigin', style=Pack(width=170))
              lblTo = toga.Label('and')
              selTo = toga.Selection(id='selTo', style=Pack(width=170))
              swtFR = toga.Switch('FR', on_toggle=frChange, is_on=True)
              swtWHR = toga.Switch('WHR', on_toggle=buildScreen, is_on=True)
              btnStart = toga.Button('Create Quiz', on_press = getImage(mainBox))
              lblResults = toga.Label('')

              # Get a list of locations to populate the select boxes with
              reqUrl = 'http://getlocations.app.php'
              rawData = requests.post(reqUrl)

              # Create a blank array to hold them
              arrLocations = []
              locArray = json.loads(rawData.content)
              for jsonLocation in locArray:
                  arrLocations.append(jsonLocation['locName'])

              # Add them to the list selections
              selFrom.items = arrLocations
              selTo.items = {'label 1': 1, 'label 2':2}

              # Add the base widgets
              mainBox.add(lblTitle)
              mainBox.add(lblIntro)
              locRow.add(lblFrom)
              locRow.add(selFrom)
              locRow.add(lblTo)
              locRow.add(selTo)
              mainBox.add(locRow)
              optRow.add(swtFR)
              optRow.add(swtWHR)
              mainBox.add(optRow)
              mainBox.add(btnStart) 
              mainBox.add(lblResults)

              # Attach this layout to the main screen
              self.main_window.content = mainBox

          # Show the main screen
          self.main_window = toga.MainWindow(title=self.formal_name)

          buildMainScreen()

          self.main_window.show()

  def main():
      return Routemaster()```

**Comment by freakboy3742** [comment]
There's a lot going on here, and it's difficult to say where exactly you've fallen off the path :-) Some of the problems are misunderstandings about Toga, and some appear to be Python misunderstandings.

The core of the problem here is that your code is the wrong "shape" - you're using a very odd combination of functions-inside functions and function invocations that... doesn't make much sense from a Python perspective. 

My main suggestion would be to go through the BeeWare and Toga tutorials. Those tutorials won't explicitly answer the question you've asked, but they will give you some general patterns of usage that should prove helpful. It would also be helpful to brush up on Python variable scoping, and how local, global, and class instance variables interact, as this seems to be the core of the problems you're having here.

However, by way of a general explainer - lets look at a single line of code:

`toga.Button('Create Quiz', on_press = getImage(mainBox))`

`getImage` is a function reference. It's a reference to a symbol that is in scope - that means it needs to be a local (which means it has been defined in the body of the function being invoked; or it's a non-local. A global is the most common form of non-local; in this case, it's from outside the scope of the function; so it's not a global (because it's bound inside a class definition); but it's still visible because of the specific combination of class and function usage that you have here.

`getImage(mainBox)` is an invocation of a function, passing `mainBox` as the first argument. This invocation happens when the code is parsed. Based on the implementation here, this means `getImage(mainBox)` will be immediately replaced with "None", because the implementation of `getImage()` doesn't do anything - so there's an implied return value of None. 

This effectively means that you're passing `on_press=None` to the `toga.Button`.

What you *should* be passing to on_press is the original function reference - i.e., `on_press=getImage`. You're telling Toga "this is the function that I want you to invoke when then button is pressed. When the button is pressed, Toga will invoke the function, and pass in the widget that was pressed as the first argument. You can't pass a "pre-invoked" function - you need the original function reference.

I'm guessing the intention here is that you want to have access to `mainBox` inside `getImage`. There are a couple of ways to do this.  
If you look at the Toga tutorials, you'll notice that all the handlers are either (a) defined inline to the method in which they are defined; or (b) defined as class instance methods. In case (1), all the widgets are available as non-local variables; in case (2), any widget that is needed is referenced as `self.widget` - because the widget is associated with the app.

I don't know if that gets you unstuck; good luck!

**Comment by montyonthebonty** [comment]
Good evening

Thanks for getting back to me - really appreciate this!

I had followed the tutorials, but as you say, I’ve struggled a little bit because most of the help online refers to more generic Python rather than Python as used in Beeware - the syntax is similar, but getting my head around the structure has been more challenging. I think I understand now what I’m aiming for and I have come up with the code below, but one thing is puzzling me. I wouldn’t expect the ‘getImage’ code to run until the button was pressed, but when I do briefcase dev, it says ‘Replacement Text’ straight away.

I’m sure it’s me not understanding something basic, but I’ve no idea what!

```
"""
Route-learning assistant
"""
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW

import requests
import json

class Routemaster(toga.App):

    lblResults = toga.Label('Initial Text')

    def startup(self):

        # Show the main screen
        self.main_window = toga.MainWindow(title=self.formal_name)

        self.buildMainScreen()

        self.main_window.show()

    # Build the main welcome screen
    def buildMainScreen(self):

        # Create the blank layout
        mainBox = toga.Box(style=Pack(direction=COLUMN, padding=20))

        # Create the base widgets
        lblTitle = toga.Label('Routemaster')
        btnCreateQuiz = toga.Button('Create', on_press = getImage(self))

        # Add the base widgets
        mainBox.add(lblTitle)
        mainBox.add(btnCreateQuiz)
        mainBox.add(self.lblResults)

        # Attach this layout to the main screen
        self.main_window.content = mainBox

def getImage(self):
    self.lblResults.text = 'Replacement Text'

def main():
    return Routemaster()

```

**Reply by freakboy3742** [reply]
Again, the same problem is occurring here. When you use `()`, you are telling Python to invoke a function. This invocation happens at when that line of code is read. So, when you say `Toga.Button(on_press=get_image())`, `get_image()` is invoked *when the Button is created*, not when the button is pressed.

`on_press` is expecting a function *reference* - and Toga will invoke the function when the button is pressed.

**Comment by montyonthebonty** [comment]
Ah, sorry - that makes sense! As you can tell, I’m used to things like JS and PHP where you put the brackets as a matter of routine!

Thanks
Chris

On 26 Aug 2022, at 02:09, Russell Keith-Magee ***@***.***> wrote:

Again, the same problem is occurring here. When you use (), you are telling Python to invoke a function. This invocation happens at when that line of code is read. So, when you say Toga.Button(on_press=get_image()), get_image() is invoked when the Button is created, not when the button is pressed.

**Reply by freakboy3742** [reply]
While "add ()" may be your habit, I'm pretty sure you'll find exactly the same thing is happening. JS and PHP don't have any magic "invoke this function later" linguistic capabilities. They just make more use of closures - functions that return functions. 

FWIW: closures are also legal in Python -
```
def my_closure():
    def my_function(widget):
        ...
    return my_function

toga.Button(on_press=my_closure())
```

Sometimes, a closure is the right answer; however, there are also often much easier ways to solve a problem.
